CIP-19: Shwap (#77)
celestiaorg/CIPs: cips/cip-shwap_protocol.md

---
title: Shwap Protocol 

> description: Shwap - the new extensible sampling protocol 

Reviewer: can the description explain what it is? Just one liner is fine. New extensible sampling protocol doesn't explain much. Something like "Shwap - a new efficient sampling protocol that combines shares and swaps" or something

Reviewer 2: +

Reviewer 2: This is resolved

author: Hlib Kanunnikov (@Wondertan)
discussions-to: https://forum.celestia.org/t/cip-shwap-protocol/1551
status: Draft
type: Standards Track
category: Data Availability, Networking
created: 2024-02-02
---

## Abstract

> This document specifies Shwap - the simple and expressive yet extensible and future-proof messaging framework aiming to

Reviewer: just started review. I would recommend if we use key words like "future-proof" to also expand on why it is future-proof later in the document. Haven't fully read doc but wanted to flag here in case.

Reviewer 2: i see no mention of "future-proof" in the doc now, so marking this resolved

solve critical inefficiencies and standardize messaging of Celestia's Data Availability p2p network.

Shwap defines a messaging framework to be exchanged around the DA p2p network in a trust-minimized way without enforcing

> transport(QUIC/TCP or IP) or application layer protocol semantics(e.g., HTTP/x). Using this framework, Shwap

Reviewer: Add space:
transport (QUIC/TCP or IP)
semantics (e.g., HTTP/x)

> transport(QUIC/TCP or IP) or application layer protocol semantics(e.g., HTTP/x). Using this framework, Shwap

Reviewer 2: ```suggestion
transport (QUIC/TCP or IP) or application layer protocol semantics (e.g., HTTP/x). Using this framework, Shwap
```

declares the most common messages and provides options for stacking them with lower-level protocols.
Shwap can be stacked together with application protocol like HTTP/x, [KadDHT][kaddht], [Bitswap][bitswap] or any custom
protocol.

## Motivation

> The current Data Availability Sampling (DAS) network protocol is inefficient. A _single_ sample operation takes log2(k)

Reviewer 2: ```suggestion
The current Data Availability Sampling (DAS) network protocol is inefficient. A _single_ sample operation takes log₂(k)
```

network roundtrips (where k is the square size). This is not practical and does not scale for the theoretically unlimited
data square that the Celestia network enables. The main motive here is a protocol with O(1) roundtrip for _multiple_
samples, preserving the assumption of having 1/n honest peers connected.

Initially, Bitswap and IPLD were adopted as the basis for the DA network protocols, including DAS,
block synchronization (BS), and blob/namespace data retrieval (ND). They gave battle-tested protocols and tooling with
pluggability to rapidly scaffold Celestia's DA network. However, it came with the price of scalability limits and
roundtrips, resulting in slower BS than block production. Before the network launch, the transition

> to the optimized [ShrEx protocol][shrex] for BS and integrating [CAR and DAGStore-based storage][storage] happened

Reviewer 2: ```suggestion
to the optimized [ShrEx protocol][shrex] for BS and integrating [CAR and DAGStore-based storage][storage] happened,
```

Author: Grammarly said to remove it for some reason, but happy to add it back

> optimizing BS and ND. However, DAS was left untouched, preserving its weak scalability and roundtrip inefficiency.

Reviewer: sentence feels incomplete or awkward between "happened" and "optimizing"

"for BS and integrating [CAR and DAGStore-based storage][storage] happened optimizing"

Author: Tried changing passive to active voice and it sounds better. WDYT?

> Before the network launch, we transitioned to the optimized [ShrEx protocol][shrex] for BS and integrated [CAR and DAGStore-based storage][storage] optimizing BS and ND.

Reviewer 2: LGTM

[nit] do we need to clarify CAR and DAGStore acryonms?

Author: The link there clarifies them, so don't think so

Shwap messaging stacked together with Bitswap protocol directly addresses described inefficiency and provides a foundation
for efficient communication for BS, ND, and beyond.

## Rationale

The atomic primitive of Celestia's DA network is the share. Shwap standardizes messaging and serialization for shares.

> Shares are grouped together, forming more complex data types(Rows, Blobs, etc.). These data types are encapsulated in

Reviewer: Add space:
types (Rows, Blobs, etc.)

> Shares are grouped together, forming more complex data types(Rows, Blobs, etc.). These data types are encapsulated in

Reviewer 2: ```suggestion
Shares are grouped together, forming more complex data types (Rows, Blobs, etc.). These data types are encapsulated in
```

> containers, e.g., Row container groups shares of a particular row. Containers can be identified with the share identifiers

Reviewer: recommend to rewrite to:
"containers. For example, row container groups..."

in order to request, advertise or index the containers. The combination of containers and identifiers provides an extensible
and expressive messaging framework for groups of shares and enables efficient single roundtrip request-response
communication.

Many share groups or containers are known in the Celestia network, and systemizing this is the main reason behind setting
up this simple messaging framework. A single place with all the possible Celestia DA messages must be defined, which node
software and protocol researchers can rely on and coordinate. Besides, this framework is designed to be

> future-proof and sustain changes in the core protocol's data structures and proving system as long shares stay the

Reviewer: Can we expand on why it is future-proof?

Commonly used terms in this document are described below.

_**Shwap**_: The protocol described by this document. Shwap is a
portmanteau name of words share and swap.

_**[Share][shares]**_: The core data structure of DataSquare **"swapped"** between peers.

_**[DataSquare][square]**_: The DA square format used by Celestia DA network.

_**[DAH][dah]**_: The Data Availability Header with Row and Column commitments.

_**[Namespace][ns]**_: The namespace grouping sets of shares.

_**Peer**_: An entity that can participate in a Shwap protocol. There are three types of peers:
client, server, and node.

_**Client**_: The Peer that requests content by content identifies over Shwap.

_**Server**_: The Peer that responds with content over Shwap.

_**Node**_: The peer that is both the client and the server.

_**Proof**_: A Merkle inclusion proof of the data in the DataSquare.

### Message Framework

This section defines Shwap's messaging framework. Every group of shares that needs to be exchanged over the network
MUST define its [share identifier](#share-identifiers) and [share container](#share-containers), as well as, follow
their described rules.

#### Share Identifiers

> Identifiers MUST have a fixed size for their fields. Subsequently, protobuf SHOULD NOT be used for CID serialization due

Author: That is a good point and connects with https://github.com/celestiaorg/celestia-node/pull/3205#discussion_r1508964497

> Identifiers MUST have a fixed size for their fields. Subsequently, protobuf SHOULD NOT be used for CID serialization due

Author: This needs a bit more thinking. I might redo this requirement completely

> Identifiers MUST have a fixed size for their fields. Subsequently, protobuf SHOULD NOT be used for CID serialization due

Author: This is now fixed

to varints and lack of fixed size arrays. Instead, identifiers use simple binary big endian serialization.

Identifiers MAY embed each other to narrow down the scope of needed shares. For example, [SampleID](#sampleid) embeds
[RowID](#rowid) as every sample lay on a particular row.

#### Share Containers

Share containers encapsulate a set of data shares with [DAH][dah] inclusion proof. Share containers are identified by
[share identifiers](#share-identifiers).

#### Versioning

If a defined share container or identifier requires an incompatible change, the new message type MAY be introduced

> suffixed with a new major version starting from v1. E.g., if the Row message needs a revision, RowV1 is created.

Reviewer 2: ```suggestion
suffixed with a new major version starting from v1. _e.g._, if the Row message needs a revision, RowV1 is created.
```

### Messages

This section defines all the supported Shwap messages, including share identifiers and containers. All the new
future messages should be described in the section.

#### RowID

RowID identifies the [Row shares container](#row-container) in a [DataSquare][square].

RowID identifiers are formatted as shown below:

```text
RowID {
    Height: u64;

>     RowIndex: u16;

Author: Not sure I understand the question.

>     RowIndex: u16;

Author: Not that I would be aware of. The u16 is chosen because it maxes at 1TB blocks, which is more than enough.

>     RowIndex: u16;

Author: cc @rootulp, do we have any row number limits defined in core?

>     RowIndex: u16;

Reviewer 3: Yea! The max row number is equivalent to  2 * the max effective square size because the row number seems applicable to the extended data square. The max effective square size is currently 64 but it could be bumped to 128 via a governance proposal. 

Ref: https://gist.github.com/rootulp/bbf10f6e9cf114816aaa994eb64b63a4

>     RowIndex: u16;

Reviewer 2: leaving this discussion open to migrate into the forum

}
```

The fields with validity rules that form RowID are:

**Height**: A uint64 representing the chain height with the data square. It MUST be bigger than zero.

**RowIndex**: An uint16 representing row index points to a particular row. It MUST not exceed the number of Row roots in
[DAH][dah].

> Serialized RowID MUST have a length of 10 bytes.

Author: Yes

> Serialized RowID MUST have a length of 10 bytes.

Author: Implied by serialized. Do you think its worth expanding on that?

> Serialized RowID MUST have a length of 10 bytes.

Reviewer 4: I'd encourage the added clarity since the note about big endian encoding is kind of buried. Alternatively have a section on Serialization and move the note on big endian to that and then you can just link to that here.

> Serialized RowID MUST have a length of 10 bytes.

Author: sgtm, clarified

> Serialized RowID MUST have a length of 10 bytes.

Reviewer 2: can you confirm this is resolved @ebuchman ?

[**SampleID**](#sampleid): A SampleID of the Sample container. It MUST follow [SampleID](#sampleid) formatting and field
validity rules.

**SampleShare**: A variable size array representing the share contained in the sample. Each share MUST follow [share
formatting and validity][shares-format] rules.

**Proof**: A [protobuf formated][nmt-pb] [NMT][nmt] proof of share inclusion. It MUST follow [NMT proof verification][nmt-verify]
and be verified against the respective root from the Row or Column axis in [DAH][dah]. The axis is defined by the ProofType field.

**ProofType**: An enum defining which axis root the Proof is coming from. It MUST be either RowProofType or ColumnProofType.

#### DataID

DataID identifies [namespace][ns] Data container of shares within a _single_ Row. That is, namespace shares spanning
over multiple Rows are identified with multiple identifiers.

DataID identifiers are formatted as shown below:

```text
DataID {
    RowID;
    Namespace;
}
```

The fields with validity rules that form DataID are:

[**RowID**](#rowid): A RowID of the namespace data. It MUST follow [RowID](#rowid) formatting and field validity rules.

> [**Namespace**][ns]: A fixed-size bytes array representing the Namespace of interest. It MUST follow [Namespace][ns]

Author: Yes. I omitted that as the linked namespace spec defines that, although it does not hurt to add that.

> [**Namespace**][ns]: A fixed-size bytes array representing the Namespace of interest. It MUST follow [Namespace][ns]

Reviewer 2: this is resolved, closing

formatting and its validity rules.

Serialized DataID MUST have a length of 39 bytes.

#### Data Container

Data containers encapsulate user-submitted data under [namespaces][ns].

Data containers are protobuf formatted using the following proto3 schema:

```protobuf
syntax = "proto3";

message Data {
    bytes data_id = 1;
    repeated bytes data_shares = 2;
    Proof data_proof = 3;
}
```

The fields with validity rules that form Data containers are:

[**DataID**](#dataid): A DataID of the Data container. It MUST follow [DataID](#dataid) formatting and field validity
rules.

**DataShares**: A two-dimensional variable size byte arrays representing left data shares of a namespace in the row.
Each share MUST follow [share formatting and validity][shares-format] rules.

**Proof**: A [protobuf formated][nmt-pb] [NMT][nmt] proof of share inclusion. It MUST follow [NMT proof verification][nmt-verify]

> and be verified against the respective root from the Row or Column axis in [DAH][dah]. The axis is defined by the ProofType field.

Author: That's an artifact from copy-pasting. There should not be any proof type. Good catch!

> and be verified against the respective root from the Row or Column axis in [DAH][dah]. The axis is defined by the ProofType field.

Reviewer 2: this is also resolved

Namespace data may span over multiple rows, in which case all the data is encapsulated in multiple containers. This is

> done

Reviewer 2: ```suggestion
done.
```

> done.

Author: That's an artifact. Seems like I accidentally removed half of the sentence in some spec iteration.

> done.

Reviewer 2: LGTM. resolving

## Protocol Compositions

This section specifies compositions of Shwap with other protocols. While Shwap is transport agnostic, there are rough
edges on the protocol integration, which every composition specification has to describe.

### Bitswap

[Bitswap][bitswap] is an application-level protocol for sharing verifiable data across peer-to-peer networks.
Bitswap operates as a dynamic want-list exchange among peers in a network. Peers continuously update and share their
want lists of desired data in real time. It is promptly fetched if at least one connected peer has the needed data.
This ongoing exchange ensures that as soon as any peer acquires the sought-after data, it can instantly share it with
those in need.

Shwap is designed to be synergetic with Bitswap, as that is the primary composition to be deployed in Celestia's DA

> network. Bitswap provides the 1/N peers guarantee and can parallelize fetching across multiple peers. Both of these properties

Reviewer 2: ```suggestion
network. Bitswap provides the 1/n peers guarantee and can parallelize fetching across multiple peers. Both of these properties
```
to standardize 1/n vs. 1/N (1/n was used above already)

terminate the old infrastructure.

## Considerations

### Security

Shwap does not change the security model of Celestia's Data Availability network and changes the underlying
protocol for data retrieval.

Essentially, the network and its codebase get simplified and require less code and infrastructure to operate. This in turn
decreases the amount of implementation vulnerabilities, DOS vectors, message amplification, and resource exhaustion attacks.
However, new bugs may be introduced, as with any new protocol.

### Protobuf Serialization

Protobuf is a widely adopted serialization format and is used within Celestia's protocols. This was quite an obvious choice
for consistency reasons, even though we could choose other more efficient and advanced formats like Cap'n Proto.

### Sampling and Reconstruction

Shwap deliberately avoids specifying sampling and reconstruction logic The sampling concerns on randomness selection and 
sample picking are out of Shwap's scope and a matter of following CIPs. Shwap only provides messaging for sampling(via
[SampleID](#sampleid) and [Sample container](#sample-container)). 

> ## Reference Implementation

Reviewer: Will there be a reference implementation for Shwap later added here?

Author: I linked below two implementations. Or do you think we should add the whole code here? That's going to be huge

Reviewer 2: IMHO I think it's best to link out. we will see what @YazzyYaz thinks and then can move forward

> ---

Reviewer 2: ```suggestion
---
cip: 18
```

---
cip: 18
title: Shwap Protocol 
description: Shwap - a new messaging framework for DA and sampling
author: Hlib Kanunnikov (@Wondertan)
discussions-to: https://forum.celestia.org/t/cip-shwap-protocol/1551
status: Draft
type: Standards Track
category: Data Availability, Networking
created: 2024-02-02
---

## Abstract

> This document specifies Shwap - the simple and expressive yet extensible messaging framework aiming to

Reviewer 3: I think the name choice should be defined as early as possible

```suggestion
This document specifies Shwap (a portmanteau of share and swap)  - the simple, expressive, and extensible messaging framework aiming to
```

Reviewer 2: agreed

solve critical inefficiencies and standardize messaging of Celestia's Data Availability p2p network.

Shwap defines a messaging framework to be exchanged around the DA p2p network in a trust-minimized way without enforcing
transport (QUIC/TCP or IP) or application layer protocol semantics (e.g., HTTP/x). Using this framework, Shwap
declares the most common messages and provides options for stacking them with lower-level protocols.
Shwap can be stacked together with application protocol like HTTP/x, [KadDHT][kaddht], [Bitswap][bitswap] or any custom
protocol.

## Motivation

The current Data Availability Sampling (DAS) network protocol is inefficient. A _single_ sample operation takes log₂(k)

> network roundtrips (where k is the square size). This is not practical and does not scale for the theoretically unlimited

Reviewer 3: [optional] 

```suggestion
network roundtrips (where k is the extended square size). This is not practical and does not scale for the theoretically unlimited
```

> data square that the Celestia network enables. The main motive here is a protocol with O(1) roundtrip for _multiple_

Reviewer 3: [nit] does Celestia theoretically enable an unlimited data square? I haven't seen that idea expressed before and the whitepaper talks about how scaling the square size up depends on a sufficient number of light nodes sampling. If Celestia enables an unlimited data square then we'd need an unlimited number of light nodes.

IMO this proposal would be more convincing if it stated something like:

```
This is not practical and does not scale as the Celestia square size is increased.
```

Author: Yes, having enough light nodes is needed, but it doesn't contradict the fact that it is theoretically unlimited. Having enough light nodes is a requirement, but not a theoretical limit.

samples, preserving the assumption of having 1/N honest peers connected possessing the data.

Initially, Bitswap and IPLD were adopted as the basis for the DA network protocols, including DAS,
block synchronization (BS), and blob/namespace data retrieval (ND). They gave battle-tested protocols and tooling with
pluggability to rapidly scaffold Celestia's DA network. However, it came with the price of scalability limits and
roundtrips, resulting in slower BS than block production. Before the network launch, we transitioned to the optimized 
[ShrEx protocol][shrex] for BS and integrated [CAR and DAGStore-based storage][storage] optimizing BS and ND. However, 
DAS was left untouched, preserving its weak scalability and roundtrip inefficiency.

Shwap messaging stacked together with Bitswap protocol directly addresses described inefficiency and provides a foundation
for efficient communication for BS, ND, and beyond.

## Rationale

The atomic primitive of Celestia's DA network is the share. Shwap standardizes messaging and serialization for shares.
Shares are grouped together, forming more complex data types (Rows, Blobs, etc.). These data types are encapsulated in

> containers, For example, row container groups shares of a particular row. Containers can be identified with the share 

Reviewer 3: ```suggestion
containers. For example, a row container groups the shares of a particular row. Containers can be identified with the share 
```

Share containers are RECOMMENDED to use protobuf (proto3) encoding, and other formats MAY be used for serialization. A 
container MAY define multiple serialization formats.

#### Share Identifiers

Share identifiers identify [share containers](#share-containers). Identifiers are not collision-resistant and there MAY 
be different identifiers referencing the same container.

Identifiers MAY embed each other to narrow down the scope of needed shares. For example, [SampleID](#sampleid) embeds
[RowID](#rowid) as every sample lay on a particular row.

##### Serialization

Share identifiers SHOULD be serialized by concatenating big-endian representations of fields in the order defined by 
their respective formatting section. Serialized identifiers SHOULD have constant and predetermined lengths in bytes.

#### Versioning

If a defined share container or identifier requires an incompatible change, the new message type MAY be introduced
suffixed with a new major version starting from v1. E.g., if the Row message needs a revision, RowV1 is created.

### Messages

This section defines all the supported Shwap messages, including share identifiers and containers. All the new

> future messages should be described in the section.

Reviewer 3: [question] in which section?

Author: This

Reviewer 3: ahh so `the` is just a typo and should be `this`

#### RowID

RowID identifies the [Row shares container](#row-container) in a [DataSquare][square].

RowID identifiers are formatted as shown below:

```text
RowID {
    Height: u64;
    RowIndex: u16;
}
```

The fields with validity rules that form RowID are:

**Height**: A uint64 representing the chain height with the data square. It MUST be bigger than zero.

**RowIndex**: An uint16 representing row index points to a particular row. The 16 bit limit fits data squares up to 2TB.

> It MUST not exceed the number of Row roots in[DAH][dah].

Reviewer 3: ```suggestion
It MUST not exceed the number of Row roots in [DAH][dah].
```

CIDs are required to have fixed and deterministic sizes. Making share identifiers compliant with CID 
prevents protobuf usage due to its reliance on varints and dynamic byte arrays serialization in.

The naive question would be: "Why not make content verification after Bitswap provided it back over its API?" Intuitively,
this would simplify much and would not require "hacking" CID. However, this has an important downside - the Bitswap, in 
such a case, would consider the request finalized and the content as fetched and valid, sending a DONT_WANT message to 
its peers. In contrast, the message might still be invalid according to the verification rules.

Bitswap still requires multihashes and CID codecs to be registered. Therefore, we provide a table for the
supported [share identifiers](#share-identifiers) with their respective multihash and CID codec codes. This table
should be extended whenever any new share identifier is added.

| Name     | Multihash | Codec  |
|----------|-----------|--------|
| RowID    | 0x7801    | 0x7800 |
| SampleID | 0x7811    | 0x7810 |
| DataID   | 0x7821    | 0x7820 |

> 

Reviewer 5: Does this also imply, that incompatible changes to CID messages would also require introduction new multihash code and codec code

Author: Yes, it's worth documenting that.

Author: Done

---
title: Shwap Protocol 
description: Shwap - the new extensible sampling protocol 
author: Hlib Kanunnikov (@Wondertan)
discussions-to: https://forum.celestia.org/t/cip-shwap-protocol/1551
status: Draft
type: Standards Track
category: Data Availability, Networking
created: 2024-02-02
---

## Abstract

> This document describes a high level overview of the Shwap p2p protocol. Shwap provides scalable and extensible 

Reviewer 3: What is the rationale for the name "Shwap"?

Author: There is no "rationale". But it is short for SharesSwap and is something that is to remember, short and sounding. It also gives me the vibe of being fast, like a snap.

Reviewer 3: Thanks! We could add something like this to the CIP to front run the same question by other readers:

> This document describes a high level overview of the Shwap p2p protocol. Shwap is a portmanteau of share and swap.

Author: Good suggestion!

framework for exchanging and swapping of shared data for Celestia's Data Availability network and beyond.

## Motivation

> Current Data Availability Sampling (DAS) network protocol is inefficient. A _single_ sample operation takes O(log2(N)) network

Reviewer 3: ```suggestion
The current Data Availability Sampling (DAS) network protocol is inefficient. A _single_ sample operation takes O(log2(N)) network
```

round-trips(where N is the square size). This is not practical and does not scale for theoretically unlimited data square

> Celestia network enables. The main motive here is a protocol with O(1) round-trip for _multiple_ samples, preserving 

Reviewer 3: [grammar]

This is not practical and does not scale for the theoretically unlimited data square that the Celestia network enables.

the assumption of having a 1/N honest peers connected.

Initially, the Bitswap and IPLD were adopted as a base for the DA network protocols, including DAS, 

> block synchronization (BS) and blob/namespace data retrieval (ND). They gave battle tested protocols and tooling with 

Reviewer 3: [grammar]

Initially, Bitswap and IPLD were adopted as the basis for the DA network protocols, including DAS, block synchronization (BS), and blob/namespace data retrieval (ND).

plugability to rapidly scaffold Celestia's DA network. However, it came with the price of scalability limits and 
round-trips resulting in BS slower than block production. Before network launch, the transition 
to the optimized [ShrEx protocol][shrex] for BS and integrating [CAR and DAGStore-based storage][storage] happened 
optimizing BS and ND. However, DAS was left untouched preserving its weak scalability and roundtrip inefficiency. Shwap 
specifically addresses these and additionally provides an extensible and flexible framework for BS, ND and beyond. 

## Specification

[Shwap Protocol Specification][spec]

## Backwards Compatibility

Shwap in incompatible with old sampling protocol. 

After rigorous investigation, celestia-node team decided against _implementing_ backwards compatibility with 
the old protocol into the node client due to immense complications it brings. Instead, the simple and time efficient 
strategy is to transiently deploy infrastructure for old and new versions, allowing network participants to gradually 
migrate to the new version. Once, we know that majority has migrated, we will terminate old infrastructure and deprecate 

> old version.

Reviewer 3: question: how can we terminate the old infrastructure at the same time as deprecating the old version? It seems like we need to deprecate the old version and commit to a future date where the old version's support will expire. After that date we can terminate old infrastructure.

Author: That's true, and that is how we wanted to approach this, good catch!

Author: Update the paragraph, check it out! And thank you for your review

## Reference Implementation

- [Go reference implementation][gimpl]
- [Rust implementation][rimpl]

## Security Considerations

Shwap does not change the security model of Celestia's Data Availability network and changes the underlying 
protocol for data retrieval.

Largely, the network and its codebase get simplified and requires less code and infrastructure to operate. This in turn 
decreases amount of implementation vulnerabilities, DOS vectors, message amplification and resource exhaustion attacks. 

## Copyright

Copyright and related rights waived via [CC0](../LICENSE).

[shrex]: https://github.com/celestiaorg/celestia-node/blob/0abd16bbb05bf3016595498844a588ef55c63d2d/docs/adr/adr-013-blocksync-overhaul-part-2.md
[storage]: https://github.com/celestiaorg/celestia-node/blob/a33c80e20da684d656c7213580be7878bcd27cf4/docs/adr/adr-011-blocksync-overhaul-part-1.md
[gimpl]: https://github.com/celestiaorg/celestia-node/pull/2675
[rimpl]: https://github.com/eigerco/lumina/blob/561640072114fa5c4ed807e94882473476a41dda/node/src/p2p/shwap.rs

> [spec]: https://www.youtube.com/watch?v=dQw4w9WgXcQ

Reviewer 3: 😆

Reviewer 2: i'm assuming this is a placeholder lol
